rfcs/2020-08-21-3092-apache-metrics-source.md

```diff
@@ -0,0 +1,246 @@
+
+The `host` key will be set to the host parsed out of the `endpoint`.
+
+## Rationale
+
+Apache HTTP Server is a fairly common webserver. If we do not support ingesting
+metrics from it, it is likely to push people to use another tool to forward
+metrics from httpd to the desired sink.
+
+As part of Vector's vision to be the "one tool" for ingesting and shipping
+observability data, it makes sense to add as many sources as possible to reduce
+the likelihood that a user will not be able to ingest metrics from their tools.
+
+## Prior Art
+
+- [Datadog collection](https://www.datadoghq.com/blog/monitor-apache-web-server-datadog/#set-up-datadogs-apache-integration)
+- [Telegraf](https://github.com/influxdata/telegraf/tree/release-1.15/plugins/inputs/apache)
+
+## Drawbacks
+
+- Additional maintenance and integration testing burden of a new source
+
+## Alternatives
+
+### Having users run telegraf and using Vector's prometheus source to scrape it
+
+We could not add the source directly to Vector and instead instruct users to run
+Telegraf and point Vector at the exposed Prometheus scrape endpoint. This would
+leverage the already supported [telegraf Apache input
+plugin](https://github.com/influxdata/telegraf/tree/release-1.15/plugins/inputs/apache)
+
+I decided against this as it would be in contrast with one of the listed
+principles of Vector:
+
+> One Tool. All Data. - One simple tool gets your logs, metrics, and traces
+> (coming soon) from A to B.
+
+[Vector
+principles](https://vector.dev/docs/about/what-is-vector/#who-should-use-vector)
+
+On the same page, it is mentioned that Vector should be a replacement for
+Telegraf.
+
+> You SHOULD use Vector to replace Logstash, Fluent*, Telegraf, Beats, or
+> similar tools.
+
+If users are already running telegraf though, they could opt for this path.
+
+### Have a generic HTTP scrape source
+
+We could model this as a generic `http_scrape` source that has a `type` (or
+`codec`?) that would determine how what type of endpoint it is scraping.
+
+I would err away from this as I think there is some risk that some "HTTP
+scrape" endpoints will need source specific configuration. We could do this
+later if this does not end up being the case and they really are all the same.
+
+One downside of this is that I think it'd be less discoverable than a
+first-class source for each type of endpoint we support scraping.
+
+## Outstanding Questions
+
+- Do we want to apply any metric labels based on the other information
+  available via the status page? I could see labeling the `url` at least.
+  Ansnswer: label with `host` and `endpoint` as described above.
+- Do we want to have one apache source able to scrape multiple endpoints?
+  Answer: yes, the config has been updated to allow multiple endpoints.
+- Are there preferences between `apache` or `httpd` for the nomenclature? I
+  feel like `apache` is more well-known though `httpd` is more accurate. Answer:
+  standardize on `apache`.
```

Comment:
Just noting, we'll want `apache_metrics` for clarity.

Reply:
I actually think that calling it `apache_metrics` rather than just `apache`, while making it clearer that the source ingests metrics, also makes it stand out against the other sources which are not suffixed with their data type. For example, we have `statsd` rather than `statsd_metrics` and `docker` rather than `docker_logs`. One could argue that `statsd` is less ambiguous, but `docker`, in my mind, could be: it could be either docker metrics or docker logs (or both).

This is actually standing out to me as something where the precedent will stick around for a while so it's probably worth having some explicit guidance for source naming. Sources could be one of, more than one of, or all of metrics logs and traces. It seems like the vision of `vector` is to treat all observability data (logs, metrics, and traces) as first-class and so it seems like we should reflect that in the component naming and terminology.

I actually kinda like Datadog's model where they just have one `apache` integration that ingests both metrics and logs from Apache. They configure it via a separate `log` section: https://github.com/DataDog/integrations-core/blob/master/apache/datadog_checks/apache/data/conf.yaml.example#L326-L347 . Given, Datadag started with metrics and moved to logs, which I think is reflected in the fact that logs is a separate subsection rather than metrics being so. For us, we might want to have metrics/log/trace configuration at the same level. For users forwarding to datadog, they may want to have similar functionality as what dd-agent currently supports.

Reply:
Good points.

> or example, we have statsd rather than statsd_metrics and docker rather than docker_logs

Then we should rename these sources to `statsd_metrics` and `docker_logs` or change our model.

> Sources could be one of, more than one of, or all of metrics logs and traces. It seems like the vision of vector is to treat all observability data (logs, metrics, and traces) as first-class and so it seems like we should reflect that in the component naming and terminology.

It's a good point. From a UX perspective, drawing lines around components based on types makes it easier to build pipelines. For example, what happens if you connected a multi-type `docker` source to `regex_parser` that only operates on logs? Setting aside the fact that we can't technically do this, I think it complicates the UX. I am very happy to discuss this change in a separate issue, and then an RFC if we deem it worthy. Would also like other people's thoughts on the matter.

Reply:
I've been going back and forth on this all afternoon. On one hand, having a single source name, docker/apache/etc, references that single point of observability. But I also started thinking "are there circumstances in which I'd only collect metrics from a service and I'd find `apache` on its own confusing?" Also, I wonder whether having differently named sources might make pipeline configurations more clear?

Reply:
> Also, I wonder whether having differently named sources might make pipeline configurations more clear?

This is the main concern for me. Combined sources may seem convenient, but how often do you want to route logs and metrics through the same pipeline?

Reply:
> This is the main concern for me. Combined sources may seem convenient, but how often do you want to route logs and metrics through the same pipeline?

It is a good question. I note that we do have a one source (`vector`), a few sinks (`vector`, `console`, `blackhole`), and a few transforms (`filter`, `lua` `remap` ([soon](https://github.com/timberio/vector/issues/3550)) that can handle both logs and metrics, but I can imagine it would be common to need to transform them differently.

I realize this is going off on quite a tangent; we could break this discussion off to a separate issue. I think it is important though.

I currently see 3 approaches:

1. Call this `apache_metrics` and move towards renaming all sources / sinks that only handle a single type to have a suffix (like `sematext_logs`). If/when we add `trace` as another type, we'd add things like `sematex_traces`.
2. Move towards one component for each "integration type' for sinks/sources and do what I describe below to control the flow of event types through the pipeline when needed.
3. Keep doing what we are doing which seems like a mix of strategies with respect to a type handling; defining some sort of "test" to determine when the type suffix should be applied.

I find the current state to be a little confusing. Some sources/sinks integrations can handle multiple metric types ('vector`, `console`) and other ones are split into `_metrics` and `_logs` (like `datadog_logs`/`datadog_metrics`). The clearest rule I can pull out for deciding is "can the integration handle pulling or pushing both logs and metrics with the same configuration? If yes, no suffix, if no, two components with `_logs` and `_metrics`". On issue with that rule is that downstream support for observaliblity types can change, making it difficult to predict if they will add `metrics` or `logs` support in the future.

I'd personally would appreciate some guidance laid out for component naming to avoid this question cropping up in the future.

As a thought experiment, maybe it is useful to imagining we add metrics support for clickhouse (https://github.com/timberio/vector/issues/3435). Where would that go? A new component (`clickhouse_metrics`)? If so, would we rename `clickhouse` to `clickhouse_logs`? Or would we add it as part of the existing `clickhouse` sink?

Expanding on approach 2:

One way we could approach that with a single source, is allow transforms and sinks to ask for only certain datatypes with something like:

```toml
[transforms.regex]
type = "regex_parser"
inputs = ["mysource.logs"]
```

Where it would only receive logs from the source and not metrics (or traces). If there was no suffix, it would simply receive both. In the case of a transform that only works with a subset of "observability types" we could error at start-up if there is an invalid pipeline.

Another way to limit the event types if you only want certain types from the source would be something like:

```toml
[sources.apache]
type = "apache"
event_types = ["log", "metric"]
```

Perhaps with syntax sugar for one type:

```toml
[sources.apache]
type = "apache.logs"
```

I think there is something attractive about one component to observe one "integration" (like apache, nginx, etc.).

Reply:
> The clearest rule I can pull out for deciding is "can the integration handle pulling or pushing both logs and metrics with the same configuration?

I agree we should have a stronger convention here and this is a good start. Right now, we handle multi-type sources by automatically inserting a type filter where necessary in front of downstream components. We should make sure that doesn't lead to confusing scenarios if we start using it more widely.

Reply:
> Right now, we handle multi-type sources by automatically inserting a type filter where necessary in front of downstream components. 

Interesting, I wasn't aware of that. I could see that be a bit confusing if it is implicit.

As for the convention, I'm ok with starting with that one, but it does mean we'll need to deprecate sources/sinks occasionally. I'm imagining, had vector existed a few years ago, before Datadog had support for logs, we may have just called the sink `datadog` and then had to deprecated it later in-lieu of `datadog_metrics` when `datadog_logs` was added.

I'm ok with just calling this `apache_metrics` for now since, prompted by this discussion, I've observed we already have a mix of suffixed and unsuffixed sinks and sources so this doesn't seem to make the situation noticeably worse and my concern about precedent is unfounded. I can create a new issue to come up with a convention for component naming. I'm still curious to hear thoughts on separate vs. integrated sources/sinks as well.

Reply:
> I can create a new issue to come up with a convention for component naming. I'm still curious to hear thoughts on separate vs. integrated sources/sinks as well.

👍 it's worth a discussion. I'm not leaning strongly in either direction, but as you can see there are scenarios we should discuss before making a decision. I like the simplicity of single, multi-type components from a documentation standpoint, but I want to make sure it doesn't over complicate the actual UX.